src/java_tools/buildjar/java/com/google/devtools/build/buildjar/javac/FormattedDiagnostic.java

```diff
@@ -134,7 +153,48 @@ public void report(Diagnostic<? extends JavaFileObject> diagnostic) {
       }
     }
 
+    @Nullable
+    private Diagnostic<? extends JavaFileObject> maybeGetJavaConfigurationDiagnostic(
+        Diagnostic<? extends JavaFileObject> diagnostic) {
+      if (!diagnostic.getKind().equals(Kind.ERROR)) {
+        return null;
+      }
+      Matcher matcher;
+      if (diagnostic.getCode().equals("compiler.err.cant.access")
+          && arguments.system() != null
+          && (matcher =
+                  INCOMPATIBLE_SYSTEM_CLASS_PATH_ERROR.matcher(diagnostic.getMessage(ENGLISH)))
+              .find()) {
+        // The output path is of the form $PRODUCT-out/$CPU-$MODE[-exec-...]/bin/...
+        boolean isForTool = arguments.classOutput().subpath(1, 2).toString().contains("-exec-");
+        // Java 8 corresponds to class file major version 52.
+        int systemClasspathVersion = Integer.parseUnsignedInt(matcher.group("version")) - 44;
+        String message =
+            String.format(
+                "[BazelJavaConfiguration] The Java %d runtime used to run javac is not recent "
+                    + "enough to compile for the Java %d runtime in %s. Either register a Java "
+                    + "toolchain with a newer java_runtime or specify a lower %s.",
+                Runtime.version().feature(),
+                systemClasspathVersion,
+                arguments.system(),
+                isForTool ? "--tool_java_runtime_version" : "--java_runtime_version");
+        return error(message);
+      }
+      return null;
+    }
+
+    private Diagnostic<? extends JavaFileObject> error(String message) {
+      return JCDiagnostic.Factory.instance(context)
+          .create(
+              DiagnosticSource.NO_SOURCE,
+              new JCDiagnostic.SimpleDiagnosticPosition(-1),
+              CompilerProperties.Errors.ProcMessager(message));
```

Comment:
As an alternative to creating a new diagnostic and including it here, would it work to do this a post-processing step when the diagnostics are printed, maybe here: https://github.com/bazelbuild/bazel/blob/8cd6c4315876748b61a30eb4b1c3b14131261e77/src/java_tools/buildjar/java/com/google/devtools/build/buildjar/BazelJavaBuilder.java#L109-L111 ?

I'm asking because that might be a way to make this less reliant on `JCDiagnostic` internals, but perhaps that wouldn't support things we want to be able to do here, so feel free to push back.

Reply:
I implemented this alternative approach.

There does seem to be a mix of `"\n"` and `println` in the code. What is the recommended way to add line endings?

Reply:
> There does seem to be a mix of `"\n"` and `println` in the code. What is the recommended way to add line endings?

In general I suppose we should be avoiding hard-coded `"\n"` and trying to support the platform's preferred line endings. (It might be worth cleaning up the existing uses of `\n`, but obviously that's not something that needs to happen in this PR.)

Want to use `%n` / `println` / etc. for the new code in this PR?

Reply:
Having worked on cross-platform caching, I would actually prefer standardized line endings across platforms:  From that perspective, every use of `println` or `%n` is an unnecessary dependency on the execution platform. Since the outputs of the Java compilation action are platform-independent, I would see value in having stdout/stderr be platform-independent as well.

What do you think about this?

Reply:
That's interesting, I hadn't considered that aspect of it. In my day job I can mostly pretend everything uses unix line endings.

In this specific case the messages are diagnostic output, which generally aren't going to affect caching, right?

I can still see the argument for standardizing on `\n` even for diagnostics, to avoid having to reason about where it's safe to use them, and make sure platform-specific line endings don't leak into output.